database/document/gc.go

```diff
@@ -0,0 +1,122 @@
+package document
+
+import (
+	"runtime"
+	"strconv"
+	"time"
+
+	"github.com/dgraph-io/badger/v3"
+	"github.com/pingcap/log"
+	"go.uber.org/zap"
+)
+
+var (
+	lastFlattenTsKey = []byte("last_flatten_ts")
+	flattenInterval  = time.Hour * 24
+)
+
+func doGCLoop(db *badger.DB, closed chan struct{}) {
+	log.Info("badger start to run value log gc loop")
+	ticker := time.NewTicker(10 * time.Minute)
+	defer func() {
+		ticker.Stop()
+		log.Info("badger stop running value log gc loop")
+	}()
+
+	// run gc when started.
+	runGC(db)
+	for {
+		select {
+		case <-ticker.C:
+			runGC(db)
+		case <-closed:
+			return
+		}
+	}
+}
+
+func runGC(db *badger.DB) {
+	tryFlattenIfNeeded(db)
+	runValueLogGC(db)
+}
+
+func runValueLogGC(db *badger.DB) {
+	defer func() {
+		r := recover()
+		if r != nil {
+			log.Error("panic when run badger value log",
+				zap.Reflect("r", r),
+				zap.Stack("stack trace"))
+		}
+	}()
```

Comment:
Why only value log gc needs recovery?

Reply:
already move this recover to `runGC` function.